File: packages/sandboxd/src/runtime/runtime_file.rs
//! File materialization for runtime client setup files.
//!
//! Setup-file specs may create or append files before processes start. This
//! module owns directory permissions, write modes, and error messages for those
//! filesystem operations.

use std::fs::{self, DirBuilder, OpenOptions};
use std::io::Write;
use std::os::unix::fs::{DirBuilderExt, OpenOptionsExt, PermissionsExt};
use std::path::Path;

use serde_json::Value;

use super::plan::{RuntimeClientSetupFile, RuntimeFileWriteMode};

const MANAGED_BLOCK_START_MARKER: &str = "<!-- MISTLE-MANAGED:START mistle-sandbox-context -->";
const MANAGED_BLOCK_END_MARKER: &str = "<!-- MISTLE-MANAGED:END mistle-sandbox-context -->";

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RuntimeFileApplyOutcome {
    Written,
    SkippedIfAbsent,
}

pub fn apply_runtime_file(
    file: &RuntimeClientSetupFile,
) -> Result<RuntimeFileApplyOutcome, String> {
    let parent_directory = Path::new(&file.path)
        .parent()
        .ok_or_else(|| format!("runtime file path {} has no parent directory", file.path))?;

    DirBuilder::new()
        .recursive(true)
        .mode(0o755)
        .create(parent_directory)
        .map_err(|error| {
            format!(
                "failed to create parent directory {}: {error}",
                parent_directory.display()
            )
        })?;

    if matches!(file.write_mode, Some(RuntimeFileWriteMode::IfAbsent))
        && Path::new(&file.path).exists()
    {
        return Ok(RuntimeFileApplyOutcome::SkippedIfAbsent);
    }

    let content = match file.write_mode {
        Some(RuntimeFileWriteMode::Merge) => merge_runtime_file(file)?,
        _ => file.content.clone(),
    };

    let mut output_file = OpenOptions::new()
        .create(true)
        .truncate(true)
        .write(true)
        .mode(file.mode)
        .open(&file.path)
        .map_err(|error| format!("failed to write file {}: {error}", file.path))?;
    output_file
        .write_all(content.as_bytes())
        .map_err(|error| format!("failed to write file {}: {error}", file.path))?;
    fs::set_permissions(&file.path, fs::Permissions::from_mode(file.mode))
        .map_err(|error| format!("failed to set file mode for {}: {error}", file.path))?;

    Ok(RuntimeFileApplyOutcome::Written)
}

fn merge_runtime_file(file: &RuntimeClientSetupFile) -> Result<String, String> {
    let existing_content = match fs::read_to_string(&file.path) {
        Ok(content) => content,
        Err(error) if error.kind() == std::io::ErrorKind::NotFound => {
            return Ok(file.content.clone());
        }
        Err(error) => return Err(format!("failed to read file {}: {error}", file.path)),
    };

    if contains_managed_block(&file.content) {
        return Ok(merge_managed_block(&existing_content, &file.content));
    }

    if let Some(merged_json) = try_merge_json_file(file, &existing_content)? {
        return Ok(merged_json);
    }

    if looks_like_toml(&file.path, &file.content) {
        return Ok(merge_toml_file(&existing_content, &file.content));
    }

    Err(format!(
        "runtime file {} uses writeMode merge, but sandboxd could not infer a supported merge format",
        file.path
    ))
}

fn contains_managed_block(content: &str) -> bool {
    content.contains(MANAGED_BLOCK_START_MARKER) && content.contains(MANAGED_BLOCK_END_MARKER)
}

fn try_merge_json_file(
    file: &RuntimeClientSetupFile,
    existing_content: &str,
) -> Result<Option<String>, String> {
    let generated_json = match serde_json::from_str::<Value>(&file.content) {
        Ok(value) => value,
        Err(error) => {
            if file.path.ends_with(".json") {
                return Err(format!(
                    "runtime file {} uses writeMode merge for JSON, but generated content is invalid JSON: {error}",
                    file.path
                ));
            }
            return Ok(None);
        }
    };

    if !generated_json.is_object() {
        return Err(format!(
            "runtime file {} uses writeMode merge for JSON, but generated content is not a JSON object",
            file.path
        ));
    }

    let mut existing_json =
        serde_json::from_str::<Value>(existing_content).map_err(|error| {
            format!(
                "runtime file {} uses writeMode merge for JSON, but existing content is invalid JSON: {error}",
                file.path
            )
        })?;
    if !existing_json.is_object() {
        return Err(format!(
            "runtime file {} uses writeMode merge for JSON, but existing content is not a JSON object",
            file.path
        ));
    }
    merge_json_value(&mut existing_json, &generated_json, None);
    serde_json::to_string_pretty(&existing_json)
        .map(ensure_trailing_newline)
        .map(Some)
        .map_err(|error| format!("failed to serialize merged JSON for {}: {error}", file.path))
}

fn merge_json_value(existing: &mut Value, generated: &Value, parent_key: Option<&str>) {
    match (existing, generated) {
        (Value::Object(existing_object), Value::Object(generated_object)) => {
            for (key, generated_value) in generated_object {
                if is_mcp_servers_key(parent_key) {
                    existing_object.insert(key.clone(), generated_value.clone());
                    continue;
                }

                match existing_object.get_mut(key) {
                    Some(existing_value) => {
                        if key == "extensions" {
                            merge_json_extensions(existing_value, generated_value);
                            continue;
                        }
                        merge_json_value(existing_value, generated_value, Some(key));
                    }
                    None => {
                        existing_object.insert(key.clone(), generated_value.clone());
                    }
                }
            }
        }
        (existing_value, generated_value) => {
            *existing_value = generated_value.clone();
        }
    }
}

fn merge_json_extensions(existing: &mut Value, generated: &Value) {
    let (Some(existing_array), Some(generated_array)) =
        (existing.as_array_mut(), generated.as_array())
    else {
        *existing = generated.clone();
        return;
    };

    for generated_item in generated_array {
        if !existing_array
            .iter()
            .any(|existing_item| existing_item == generated_item)
        {
            existing_array.push(generated_item.clone());
        }
    }
}

fn is_mcp_servers_key(key: Option<&str>) -> bool {
    matches!(key, Some("mcp" | "mcpServers" | "mcp_servers"))
}

fn merge_managed_block(existing_content: &str, replacement_content: &str) -> String {
    if let Some(start_index) = existing_content.find(MANAGED_BLOCK_START_MARKER) {
        let search_start = start_index + MANAGED_BLOCK_START_MARKER.len();
        if let Some(relative_end_index) =
            existing_content[search_start..].find(MANAGED_BLOCK_END_MARKER)
        {
            let end_index = search_start + relative_end_index + MANAGED_BLOCK_END_MARKER.len();
            let mut merged = String::new();
            merged.push_str(&existing_content[..start_index]);
            merged.push_str(trim_trailing_newlines(replacement_content));
            merged.push_str(&existing_content[end_index..]);
            return ensure_trailing_newline(merged);
        }
    }

    let mut merged = ensure_trailing_newline(existing_content.to_owned());
    if !merged.ends_with("\n\n") {
        merged.push('\n');
    }
    merged.push_str(trim_trailing_newlines(replacement_content));
    merged.push('\n');
    merged
}

fn looks_like_toml(path: &str, content: &str) -> bool {
    path.ends_with(".toml")
        || split_lines_with_endings(content)
            .iter()
            .any(|line| toml_section_name(line).is_some())
}

fn merge_toml_file(existing_content: &str, generated_content: &str) -> String {
    let mut merged = existing_content.to_owned();
    for line in generated_toml_root_key_lines(generated_content) {
        if let Some(key) = toml_key_name(&line) {
            merged = ensure_toml_root_key(&merged, key, trim_trailing_newlines(&line));
        }
    }

    for section_name in generated_toml_section_names(generated_content) {
        let Some(section) = extract_toml_section(generated_content, &section_name) else {
            continue;
        };
        if is_mcp_server_section(&section_name) {
            merged = replace_or_append_toml_section(&merged, &section_name, &section);
            continue;
        }

        let section_lines = generated_toml_section_key_lines(&section);
        if section_lines.is_empty() {
            merged = replace_or_append_toml_section(&merged, &section_name, &section);
            continue;
        }

        for line in section_lines {
            if let Some(key) = toml_key_name(&line) {
                merged = ensure_toml_section_key(
                    &merged,
                    &section_name,
                    key,
                    trim_trailing_newlines(&line),
                );
            }
        }
    }

    ensure_trailing_newline(merged)
}

fn generated_toml_root_key_lines(content: &str) -> Vec<String> {
    split_lines_with_endings(content)
        .into_iter()
        .take_while(|line| toml_section_name(line).is_none())
        .filter(|line| toml_key_name(line).is_some())
        .collect()
}

fn generated_toml_section_names(content: &str) -> Vec<String> {
    let mut section_names = Vec::new();
    for line in split_lines_with_endings(content) {
        let Some(section_name) = toml_section_name(&line) else {
            continue;
        };
        if !section_names
            .iter()
            .any(|existing: &String| existing == section_name)
        {
            section_names.push(section_name.to_owned());
        }
    }

    section_names
}

fn generated_toml_section_key_lines(section: &str) -> Vec<String> {
    split_lines_with_endings(section)
        .into_iter()
        .filter(|line| toml_key_name(line).is_some())
        .collect()
}

fn is_mcp_server_section(section_name: &str) -> bool {
    section_name.starts_with("mcp_servers.") || section_name.starts_with("mcpServers.")
}

fn extract_toml_section(content: &str, section_name: &str) -> Option<String> {
    let lines = split_lines_with_endings(content);
    let start_index = lines
        .iter()
        .position(|line| toml_section_name(line) == Some(section_name))?;
    let end_index = lines[start_index + 1..]
        .iter()
        .position(|line| toml_section_name(line).is_some())
        .map(|offset| start_index + 1 + offset)
        .unwrap_or(lines.len());

    Some(ensure_trailing_newline(
        lines[start_index..end_index].concat(),
    ))
}

fn ensure_toml_section_key(content: &str, section_name: &str, key: &str, line: &str) -> String {
    let mut lines = split_lines_with_endings(content);
    let Some(start_index) = lines
        .iter()
        .position(|candidate| toml_section_name(candidate) == Some(section_name))
    else {
        let mut output = ensure_trailing_newline(content.to_owned());
        if !output.ends_with("\n\n") {
            output.push('\n');
        }
        output.push('[');
        output.push_str(section_name);
        output.push_str("]\n");
        output.push_str(line);
        output.push('\n');
        return output;
    };

    let end_index = lines[start_index + 1..]
        .iter()
        .position(|candidate| toml_section_name(candidate).is_some())
        .map(|offset| start_index + 1 + offset)
        .unwrap_or(lines.len());

    if let Some(relative_key_index) = lines[start_index + 1..end_index]
        .iter()
        .position(|candidate| toml_key_name(candidate) == Some(key))
    {
        lines[start_index + 1 + relative_key_index] = format!("{line}\n");
        return lines.concat();
    }

    lines.insert(end_index, format!("{line}\n"));
    lines.concat()
}

fn ensure_toml_root_key(content: &str, key: &str, line: &str) -> String {
    let mut lines = split_lines_with_endings(content);
    let first_section_index = lines
        .iter()
        .position(|candidate| toml_section_name(candidate).is_some())
        .unwrap_or(lines.len());

    if let Some(existing_key_index) = lines[..first_section_index]
        .iter()
        .position(|candidate| toml_key_name(candidate) == Some(key))
    {
        lines[existing_key_index] = format!("{line}\n");
        return lines.concat();
    }

    lines.insert(first_section_index, format!("{line}\n"));
    lines.concat()
}

fn replace_or_append_toml_section(content: &str, section_name: &str, replacement: &str) -> String {
    let mut lines = split_lines_with_endings(content);
    let Some(start_index) = lines
        .iter()
        .position(|candidate| toml_section_name(candidate) == Some(section_name))
    else {
        let mut output = ensure_trailing_newline(content.to_owned());
        if !output.ends_with("\n\n") {
            output.push('\n');
        }
        output.push_str(trim_trailing_newlines(replacement));
        output.push('\n');
        return output;
    };

    let end_index = lines[start_index + 1..]
        .iter()
        .position(|candidate| toml_section_name(candidate).is_some())
        .map(|offset| start_index + 1 + offset)
        .unwrap_or(lines.len());

    lines.splice(
        start_index..end_index,
        split_lines_with_endings(replacement),
    );
    lines.concat()
}

fn split_lines_with_endings(content: &str) -> Vec<String> {
    content.lines().map(|line| format!("{line}\n")).collect()
}

fn toml_section_name(line: &str) -> Option<&str> {
    let trimmed = line.trim();
    if !trimmed.starts_with('[') || !trimmed.ends_with(']') || trimmed.starts_with("[[") {
        return None;
    }

    Some(trimmed.trim_start_matches('[').trim_end_matches(']').trim())
}

fn toml_key_name(line: &str) -> Option<&str> {
    let trimmed = line.trim_start();
    if trimmed.starts_with('#') || trimmed.starts_with('[') {
        return None;
    }

    let (key, _) = trimmed.split_once('=')?;
    Some(key.trim())
}

fn trim_trailing_newlines(content: &str) -> &str {
    content.trim_end_matches(['\r', '\n'])
}

fn ensure_trailing_newline(mut content: String) -> String {
    if !content.ends_with('\n') {
        content.push('\n');
    }

    content
}

File: packages/sandboxd/runtime/runtime_file.go
package runtime

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

const (
	ManagedBlockStartMarker = "<!-- MISTLE-MANAGED:START mistle-sandbox-context -->"
	ManagedBlockEndMarker   = "<!-- MISTLE-MANAGED:END mistle-sandbox-context -->"
)

type RuntimeFileApplyOutcome string

const (
	RuntimeFileApplyOutcomeWritten         RuntimeFileApplyOutcome = "written"
	RuntimeFileApplyOutcomeSkippedIfAbsent RuntimeFileApplyOutcome = "skipped_if_absent"
)

func ApplyRuntimeFile(file RuntimeClientSetupFile) (RuntimeFileApplyOutcome, error) {
	parentDirectory := filepath.Dir(file.Path)
	if parentDirectory == "." || parentDirectory == "" {
		return "", fmt.Errorf("runtime file path %s has no parent directory", file.Path)
	}
	if err := os.MkdirAll(parentDirectory, 0o755); err != nil {
		return "", fmt.Errorf("failed to create parent directory %s: %w", parentDirectory, err)
	}

	if file.WriteMode != nil && *file.WriteMode == RuntimeFileWriteModeIfAbsent {
		if _, err := os.Stat(file.Path); err == nil {
			return RuntimeFileApplyOutcomeSkippedIfAbsent, nil
		} else if !os.IsNotExist(err) {
			return "", fmt.Errorf("failed to inspect file %s: %w", file.Path, err)
		}
	}

	content := file.Content
	if file.WriteMode != nil && *file.WriteMode == RuntimeFileWriteModeMerge {
		mergedContent, err := mergeRuntimeFile(file)
		if err != nil {
			return "", err
		}
		content = mergedContent
	}

	if err := os.WriteFile(file.Path, []byte(content), os.FileMode(file.Mode)); err != nil {
		return "", fmt.Errorf("failed to write file %s: %w", file.Path, err)
	}
	if err := os.Chmod(file.Path, os.FileMode(file.Mode)); err != nil {
		return "", fmt.Errorf("failed to set file mode for %s: %w", file.Path, err)
	}
	return RuntimeFileApplyOutcomeWritten, nil
}

func mergeRuntimeFile(file RuntimeClientSetupFile) (string, error) {
	existingContentBytes, err := os.ReadFile(file.Path)
	if err != nil {
		if os.IsNotExist(err) {
			return file.Content, nil
		}
		return "", fmt.Errorf("failed to read file %s: %w", file.Path, err)
	}
	existingContent := string(existingContentBytes)

	if containsManagedBlock(file.Content) {
		return mergeManagedBlock(existingContent, file.Content), nil
	}
	if mergedJSON, ok, err := tryMergeJSONFile(file, existingContent); err != nil || ok {
		return mergedJSON, err
	}
	if looksLikeTOML(file.Path, file.Content) {
		return mergeTOMLFile(existingContent, file.Content), nil
	}
	return "", fmt.Errorf(
		"runtime file %s uses writeMode merge, but sandboxd could not infer a supported merge format",
		file.Path,
	)
}

func containsManagedBlock(content string) bool {
	return strings.Contains(content, ManagedBlockStartMarker) && strings.Contains(content, ManagedBlockEndMarker)
}

func tryMergeJSONFile(file RuntimeClientSetupFile, existingContent string) (string, bool, error) {
	var generatedJSON any
	if err := json.Unmarshal([]byte(file.Content), &generatedJSON); err != nil {
		if strings.HasSuffix(file.Path, ".json") {
			return "", false, fmt.Errorf(
				"runtime file %s uses writeMode merge for JSON, but generated content is invalid JSON: %w",
				file.Path,
				err,
			)
		}
		return "", false, nil
	}
	generatedObject, ok := generatedJSON.(map[string]any)
	if !ok {
		return "", false, fmt.Errorf(
			"runtime file %s uses writeMode merge for JSON, but generated content is not a JSON object",
			file.Path,
		)
	}

	var existingJSON any
	if err := json.Unmarshal([]byte(existingContent), &existingJSON); err != nil {
		return "", false, fmt.Errorf(
			"runtime file %s uses writeMode merge for JSON, but existing content is invalid JSON: %w",
			file.Path,
			err,
		)
	}
	existingObject, ok := existingJSON.(map[string]any)
	if !ok {
		return "", false, fmt.Errorf(
			"runtime file %s uses writeMode merge for JSON, but existing content is not a JSON object",
			file.Path,
		)
	}
	mergeJSONObject(existingObject, generatedObject, "")
	mergedJSON, err := json.MarshalIndent(existingObject, "", "  ")
	if err != nil {
		return "", false, fmt.Errorf("failed to serialize merged JSON for %s: %w", file.Path, err)
	}
	return ensureTrailingNewline(string(mergedJSON)), true, nil
}

func mergeJSONObject(existing map[string]any, generated map[string]any, parentKey string) {
	for key, generatedValue := range generated {
		if isMCPServersKey(parentKey) {
			existing[key] = generatedValue
			continue
		}
		existingValue, ok := existing[key]
		if !ok {
			existing[key] = generatedValue
			continue
		}
		if key == "extensions" {
			existing[key] = mergeJSONArray(existingValue, generatedValue)
			continue
		}
		existingChild, existingOK := existingValue.(map[string]any)
		generatedChild, generatedOK := generatedValue.(map[string]any)
		if existingOK && generatedOK {
			mergeJSONObject(existingChild, generatedChild, key)
			continue
		}
		existing[key] = generatedValue
	}
}

func mergeJSONArray(existingValue any, generatedValue any) any {
	existingArray, existingOK := existingValue.([]any)
	generatedArray, generatedOK := generatedValue.([]any)
	if !existingOK || !generatedOK {
		return generatedValue
	}
	for _, generatedItem := range generatedArray {
		if !jsonArrayContains(existingArray, generatedItem) {
			existingArray = append(existingArray, generatedItem)
		}
	}
	return existingArray
}

func jsonArrayContains(items []any, target any) bool {
	targetBytes, targetErr := json.Marshal(target)
	if targetErr != nil {
		return false
	}
	for _, item := range items {
		itemBytes, itemErr := json.Marshal(item)
		if itemErr == nil && string(itemBytes) == string(targetBytes) {
			return true
		}
	}
	return false
}

func isMCPServersKey(key string) bool {
	return key == "mcp" || key == "mcpServers" || key == "mcp_servers"
}

func mergeManagedBlock(existingContent string, replacementContent string) string {
	startIndex := strings.Index(existingContent, ManagedBlockStartMarker)
	if startIndex >= 0 {
		searchStart := startIndex + len(ManagedBlockStartMarker)
		relativeEndIndex := strings.Index(existingContent[searchStart:], ManagedBlockEndMarker)
		if relativeEndIndex >= 0 {
			endIndex := searchStart + relativeEndIndex + len(ManagedBlockEndMarker)
			return ensureTrailingNewline(existingContent[:startIndex] + trimTrailingNewlines(replacementContent) + existingContent[endIndex:])
		}
	}
	merged := ensureTrailingNewline(existingContent)
	if !strings.HasSuffix(merged, "\n\n") {
		merged += "\n"
	}
	return merged + trimTrailingNewlines(replacementContent) + "\n"
}

func looksLikeTOML(path string, content string) bool {
	if strings.HasSuffix(path, ".toml") {
		return true
	}
	for _, line := range splitLinesWithEndings(content) {
		if tomlSectionName(line) != "" {
			return true
		}
	}
	return false
}

func mergeTOMLFile(existingContent string, generatedContent string) string {
	merged := existingContent
	for _, line := range generatedTOMLRootKeyLines(generatedContent) {
		key := tomlKeyName(line)
		if key != "" {
			merged = ensureTOMLRootKey(merged, key, trimTrailingNewlines(line))
		}
	}
	for _, sectionName := range generatedTOMLSectionNames(generatedContent) {
		section := extractTOMLSection(generatedContent, sectionName)
		if section == "" {
			continue
		}
		if isMCPServerSection(sectionName) {
			merged = replaceOrAppendTOMLSection(merged, sectionName, section)
			continue
		}
		sectionLines := generatedTOMLSectionKeyLines(section)
		if len(sectionLines) == 0 {
			merged = replaceOrAppendTOMLSection(merged, sectionName, section)
			continue
		}
		for _, line := range sectionLines {
			key := tomlKeyName(line)
			if key != "" {
				merged = ensureTOMLSectionKey(merged, sectionName, key, trimTrailingNewlines(line))
			}
		}
	}
	return ensureTrailingNewline(merged)
}

func generatedTOMLRootKeyLines(content string) []string {
	lines := []string{}
	for _, line := range splitLinesWithEndings(content) {
		if tomlSectionName(line) != "" {
			break
		}
		if tomlKeyName(line) != "" {
			lines = append(lines, line)
		}
	}
	return lines
}

func generatedTOMLSectionNames(content string) []string {
	sectionNames := []string{}
	seen := map[string]bool{}
	for _, line := range splitLinesWithEndings(content) {
		sectionName := tomlSectionName(line)
		if sectionName != "" && !seen[sectionName] {
			seen[sectionName] = true
			sectionNames = append(sectionNames, sectionName)
		}
	}
	return sectionNames
}

func generatedTOMLSectionKeyLines(section string) []string {
	lines := []string{}
	for _, line := range splitLinesWithEndings(section) {
		if tomlKeyName(line) != "" {
			lines = append(lines, line)
		}
	}
	return lines
}

func isMCPServerSection(sectionName string) bool {
	return strings.HasPrefix(sectionName, "mcp_servers.") || strings.HasPrefix(sectionName, "mcpServers.")
}

func extractTOMLSection(content string, sectionName string) string {
	lines := splitLinesWithEndings(content)
	startIndex := -1
	for index, line := range lines {
		if tomlSectionName(line) == sectionName {
			startIndex = index
			break
		}
	}
	if startIndex < 0 {
		return ""
	}
	endIndex := len(lines)
	for index := startIndex + 1; index < len(lines); index++ {
		if tomlSectionName(lines[index]) != "" {
			endIndex = index
			break
		}
	}
	return ensureTrailingNewline(strings.Join(lines[startIndex:endIndex], ""))
}

func ensureTOMLSectionKey(content string, sectionName string, key string, line string) string {
	lines := splitLinesWithEndings(content)
	startIndex := -1
	for index, candidate := range lines {
		if tomlSectionName(candidate) == sectionName {
			startIndex = index
			break
		}
	}
	if startIndex < 0 {
		output := ensureTrailingNewline(content)
		if !strings.HasSuffix(output, "\n\n") {
			output += "\n"
		}
		return output + "[" + sectionName + "]\n" + line + "\n"
	}
	endIndex := nextTOMLSectionIndex(lines, startIndex+1)
	for index := startIndex + 1; index < endIndex; index++ {
		if tomlKeyName(lines[index]) == key {
			lines[index] = line + "\n"
			return strings.Join(lines, "")
		}
	}
	lines = append(lines[:endIndex], append([]string{line + "\n"}, lines[endIndex:]...)...)
	return strings.Join(lines, "")
}

func ensureTOMLRootKey(content string, key string, line string) string {
	lines := splitLinesWithEndings(content)
	firstSectionIndex := nextTOMLSectionIndex(lines, 0)
	for index := 0; index < firstSectionIndex; index++ {
		if tomlKeyName(lines[index]) == key {
			lines[index] = line + "\n"
			return strings.Join(lines, "")
		}
	}
	lines = append(lines[:firstSectionIndex], append([]string{line + "\n"}, lines[firstSectionIndex:]...)...)
	return strings.Join(lines, "")
}

func replaceOrAppendTOMLSection(content string, sectionName string, replacement string) string {
	lines := splitLinesWithEndings(content)
	startIndex := -1
	for index, candidate := range lines {
		if tomlSectionName(candidate) == sectionName {
			startIndex = index
			break
		}
	}
	if startIndex < 0 {
		output := ensureTrailingNewline(content)
		if !strings.HasSuffix(output, "\n\n") {
			output += "\n"
		}
		return output + trimTrailingNewlines(replacement) + "\n"
	}
	endIndex := nextTOMLSectionIndex(lines, startIndex+1)
	replacementLines := splitLinesWithEndings(replacement)
	lines = append(lines[:startIndex], append(replacementLines, lines[endIndex:]...)...)
	return strings.Join(lines, "")
}

func nextTOMLSectionIndex(lines []string, startIndex int) int {
	for index := startIndex; index < len(lines); index++ {
		if tomlSectionName(lines[index]) != "" {
			return index
		}
	}
	return len(lines)
}

func splitLinesWithEndings(content string) []string {
	if content == "" {
		return nil
	}
	sourceLines := strings.SplitAfter(content, "\n")
	if sourceLines[len(sourceLines)-1] == "" {
		sourceLines = sourceLines[:len(sourceLines)-1]
	}
	lines := make([]string, 0, len(sourceLines))
	for _, line := range sourceLines {
		if strings.HasSuffix(line, "\n") {
			lines = append(lines, line)
		} else {
			lines = append(lines, line+"\n")
		}
	}
	return lines
}

func tomlSectionName(line string) string {
	trimmed := strings.TrimSpace(line)
	if !strings.HasPrefix(trimmed, "[") || !strings.HasSuffix(trimmed, "]") || strings.HasPrefix(trimmed, "[[") {
		return ""
	}
	return strings.TrimSpace(strings.TrimSuffix(strings.TrimPrefix(trimmed, "["), "]"))
}

func tomlKeyName(line string) string {
	trimmed := strings.TrimLeft(line, " \t")
	if strings.HasPrefix(trimmed, "#") || strings.HasPrefix(trimmed, "[") {
		return ""
	}
	key, _, ok := strings.Cut(trimmed, "=")
	if !ok {
		return ""
	}
	return strings.TrimSpace(key)
}

func trimTrailingNewlines(content string) string {
	return strings.TrimRight(content, "\r\n")
}

func ensureTrailingNewline(content string) string {
	if strings.HasSuffix(content, "\n") {
		return content
	}
	return content + "\n"
}
